File: ios/Packages/Domain/Sources/Domain/Credits.swift
// Credit arithmetic (PLAN §7.2 "Credit metering (two buckets)", §9.2). Pure, unit-tested; the server ledger applies the same rules in a transaction.
import Foundation

public enum CreditBucket: String, Codable, Sendable { case grant, purchased, mixed }

public struct CreditBalances: Codable, Equatable, Sendable {
    public var grant: Int
    public var grantExpiresAt: Date?
    public var purchased: Int

    public init(grant: Int, grantExpiresAt: Date? = nil, purchased: Int) {
        self.grant = grant; self.grantExpiresAt = grantExpiresAt; self.purchased = purchased
    }

    /// Expired grants count as zero (the purge job zeroes them; the client must not display or spend them).
    public func effectiveGrant(at now: Date) -> Int {
        if let exp = grantExpiresAt, exp <= now { return 0 }
        return max(0, grant)
    }

    /// Purchased credits can go negative after an Apple refund/revoke; that blocks generation until topped up.
    public func canGenerate(at now: Date) -> Bool { purchased >= 0 && (effectiveGrant(at: now) + purchased) > 0 }
    public func total(at now: Date) -> Int { effectiveGrant(at: now) + purchased }
}

public struct CreditDebit: Equatable, Sendable {
    public var balances: CreditBalances
    public var bucket: CreditBucket
    public var grantUsed: Int
    public var purchasedUsed: Int
}

public enum CreditError: Error, Equatable, Sendable {
    case insufficient(needed: Int, available: Int)
    case purchasedNegative
    case dailyCapReached(resetsAt: Date)
}

public enum Credits {
    /// One user-facing Look = 5 internal credits (§9.2). Prices come from Remote Config in production; these are the client defaults.
    public static let creditsPerLook = 5
    public static let dailyHardCapCredits = 300 // = 60 Looks, independent of balance (§7.2)

    public enum Metered: Sendable {
        case lookStandard, lookHD, twin, glowUp
        public var credits: Int {
            switch self {
            case .lookStandard, .glowUp: return Credits.creditsPerLook
            case .lookHD, .twin: return 2 * Credits.creditsPerLook
            }
        }
    }

    public static func looks(fromCredits credits: Int) -> Int { credits >= 0 ? credits / creditsPerLook : -((-credits + creditsPerLook - 1) / creditsPerLook) }
    public static func credits(forLooks looks: Int) -> Int { looks * creditsPerLook }

    /// Debit order grant → purchased (§7.2). Fails without touching balances when the total is short or purchased is negative.
    public static func debit(_ amount: Int, from balances: CreditBalances, now: Date, dailyUsed: Int = 0, dailyResetAt: Date? = nil) throws -> CreditDebit {
        precondition(amount > 0, "debit amount must be positive")
        if balances.purchased < 0 { throw CreditError.purchasedNegative }
        if dailyUsed + amount > dailyHardCapCredits { throw CreditError.dailyCapReached(resetsAt: dailyResetAt ?? now) }
        let grant = balances.effectiveGrant(at: now)
        let available = grant + balances.purchased
        if available < amount { throw CreditError.insufficient(needed: amount, available: available) }
        let fromGrant = min(grant, amount)
        let fromPurchased = amount - fromGrant
        var next = balances
        next.grant = grant - fromGrant
        next.purchased -= fromPurchased
        let bucket: CreditBucket = fromPurchased == 0 ? .grant : (fromGrant == 0 ? .purchased : .mixed)
        return CreditDebit(balances: next, bucket: bucket, grantUsed: fromGrant, purchasedUsed: fromPurchased)
    }

    /// Refund on terminal failure in the same shape and buckets as the debit (§7.2).
    public static func refund(_ debit: CreditDebit, to balances: CreditBalances) -> CreditBalances {
        var next = balances
        next.grant += debit.grantUsed
        next.purchased += debit.purchasedUsed
        return next
    }

    /// Apple refund/revoke hits purchased only and may go negative (§7.2, §9.2).
    public static func revoke(purchased amount: Int, from balances: CreditBalances) -> CreditBalances {
        var next = balances
        next.purchased -= amount
        return next
    }

    /// Monthly Plus grant: rolls over at most one month, i.e. the new balance is the new grant plus whatever is still unexpired (§9.2).
    public static func renewGrant(_ balances: CreditBalances, grant: Int, now: Date, expiresAt: Date) -> CreditBalances {
        var next = balances
        next.grant = balances.effectiveGrant(at: now) + grant
        next.grantExpiresAt = expiresAt
        return next
    }
}

/// "12 Looks from your plan (renews 3 Oct) · 20 purchased" — the only denomination the user ever sees (§9.2).
public struct LooksDisplay: Equatable, Sendable {
    public let planLooks: Int
    public let purchasedLooks: Int
    public let renewsAt: Date?

    public init(_ balances: CreditBalances, now: Date) {
        planLooks = Credits.looks(fromCredits: balances.effectiveGrant(at: now))
        purchasedLooks = Credits.looks(fromCredits: balances.purchased)
        renewsAt = balances.grantExpiresAt
    }
}

File: backend/src/ledger/credits.ts
// Credit arithmetic (PLAN §7.2, §9.2). Pure; the Firestore transaction in jobs/createJob.ts applies it. Mirrors ios/Packages/Domain/Credits.swift.

export type CreditBucket = "grant" | "purchased" | "mixed";

export type CreditBalances = {
  grant: number;
  grant_expires_at: number | null; // epoch ms
  purchased: number;
};

export type CreditDebit = {
  balances: CreditBalances;
  bucket: CreditBucket;
  grant_used: number;
  purchased_used: number;
};

export class CreditError extends Error {
  constructor(
    public readonly code: "insufficient" | "purchased_negative" | "daily_cap_reached",
    message: string,
    public readonly details: Record<string, number> = {},
  ) {
    super(message);
  }
}

export const CREDITS_PER_LOOK = 5;
export const DAILY_HARD_CAP_CREDITS = 300; // = 60 Looks, independent of balance (§7.2)

/** Default credit prices per metered job type (§9.2); production reads the same table from Remote Config. */
export const DEFAULT_CREDIT_PRICES: Record<string, number> = {
  look: CREDITS_PER_LOOK,
  look_hd: 2 * CREDITS_PER_LOOK,
  twin: 2 * CREDITS_PER_LOOK,
  glowup: CREDITS_PER_LOOK,
};

export function effectiveGrant(b: CreditBalances, nowMs: number): number {
  if (b.grant_expires_at !== null && b.grant_expires_at <= nowMs) return 0;
  return Math.max(0, b.grant);
}

export function looksFromCredits(credits: number): number {
  return credits >= 0 ? Math.floor(credits / CREDITS_PER_LOOK) : -Math.ceil(-credits / CREDITS_PER_LOOK);
}

export function canGenerate(b: CreditBalances, nowMs: number): boolean {
  return b.purchased >= 0 && effectiveGrant(b, nowMs) + b.purchased > 0;
}

/** Debit order grant → purchased. Throws without changing anything when short, when purchased is negative, or over the daily cap. */
export function debit(amount: number, b: CreditBalances, nowMs: number, dailyUsed = 0): CreditDebit {
  if (!(amount > 0)) throw new Error("debit amount must be positive");
  if (b.purchased < 0) throw new CreditError("purchased_negative", "A refund left your purchased Looks negative; top up to continue.");
  if (dailyUsed + amount > DAILY_HARD_CAP_CREDITS) {
    throw new CreditError("daily_cap_reached", "You've used today's Looks — resets at 00:00", { daily_used: dailyUsed, cap: DAILY_HARD_CAP_CREDITS });
  }
  const grant = effectiveGrant(b, nowMs);
  const available = grant + b.purchased;
  if (available < amount) throw new CreditError("insufficient", "Not enough Looks.", { needed: amount, available });
  const fromGrant = Math.min(grant, amount);
  const fromPurchased = amount - fromGrant;
  const bucket: CreditBucket = fromPurchased === 0 ? "grant" : fromGrant === 0 ? "purchased" : "mixed";
  return {
    balances: { grant: grant - fromGrant, grant_expires_at: b.grant_expires_at, purchased: b.purchased - fromPurchased },
    bucket,
    grant_used: fromGrant,
    purchased_used: fromPurchased,
  };
}

/** Refund on terminal failure in the same shape and buckets (§7.2). */
export function refund(d: CreditDebit, b: CreditBalances): CreditBalances {
  return { ...b, grant: b.grant + d.grant_used, purchased: b.purchased + d.purchased_used };
}

/** Apple refund/revoke hits purchased only and may go negative (§7.2). */
export function revoke(amount: number, b: CreditBalances): CreditBalances {
  return { ...b, purchased: b.purchased - amount };
}

/** Monthly grant: whatever is still unexpired rolls over (≤ 1 month by construction) and the new grant is added (§9.2). */
export function renewGrant(b: CreditBalances, grant: number, nowMs: number, expiresAtMs: number): CreditBalances {
  return { grant: effectiveGrant(b, nowMs) + grant, grant_expires_at: expiresAtMs, purchased: b.purchased };
}
